### PR TITLE
[api] eth_getBlobSidecar

### PR DESCRIPTION
**Pull request by envestcc**
# Description
introduce new api `eth_getBlobSidecars` to return blobs in a block.

## Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/coreservice.go [hunk 1]
```diff
@@ -1230,6 +1233,31 @@ func (core *coreService) getBlockByHeight(height uint64) (*apitypes.BlockWithRec
...
+			BlockNumber: height,
+			BlockHash:   common.BytesToHash(blkHash[:]),
+			TxIndex:     txIndexes[i],
+			TxHash:      common.BytesToHash(txHashes[i][:]),
+		})
+	}
+	return res, nil
+}
+
+func (core *coreService) getBlobSidecars(height uint64) (blobs []any, txIndexes []uint64, txHashes []hash.Hash256) {
```
Comment (dustinxie) [inline api/coreservice.go:1256]: this func should only return `blobs`
`txIndexes, txHashes` should come from core's other API funcs

Comment (dustinxie) [reply api/coreservice.go:1256]: also txIndexes --> txIndices

Comment (envestcc) [reply api/coreservice.go]: fixed

File: api/coreservice.go [hunk 2]
```diff
@@ -1213,6 +1217,58 @@ func (core *coreService) getBlockByHeight(height uint64) (*apitypes.BlockWithRec
 	}, nil
 }
 
+func (core *coreService) BlobSidecarsByHeight(height uint64) ([]*apitypes.BlobSidecarResult, error) {
+	header, err := core.bc.BlockHeaderByHeight(height)
```
Comment (dustinxie) [inline api/coreservice.go:1221]: move to L1238

File: api/coreservice.go [hunk 3]
```diff
@@ -1213,6 +1217,58 @@ func (core *coreService) getBlockByHeight(height uint64) (*apitypes.BlockWithRec
...
+	txIndices := make([]uint64, 0)
+	for _, txHash := range txHashes {
+		_, _, index, err := core.ActionByActionHash(txHash)
+		if err != nil {
+			return nil, err
+		}
+		txIndices = append(txIndices, uint64(index))
+	}
+	blkHash := header.HashBlock()
+	for i, blob := range blobs {
```
Comment (dustinxie) [inline api/coreservice.go:1239]: move L1232 to here, get the `index` and use it at L1244, then we don't need to define `txIndices`?

File: api/coreservice.go [hunk 4]
```diff
@@ -1213,6 +1217,58 @@ func (core *coreService) getBlockByHeight(height uint64) (*apitypes.BlockWithRec
...
+		res = append(res, &apitypes.BlobSidecarResult{
+			BlobSidecar: blob,
+			BlockNumber: height,
+			BlockHash:   common.BytesToHash(blkHash[:]),
```
Comment (dustinxie) [inline api/coreservice.go:1243]: `BlockNumber, BlockHash` is common for all tx, only need 1 copy? i mean like below?
```
apitypes.BlobSidecarResult {
    BlockNumber
    BlockHash
    [] struct {
        blobSidecar
        txIndex
        txHash
    }
}
```

File: api/web3server.go [hunk 5]
```diff
@@ -982,6 +984,26 @@ func (svr *web3Handler) unsubscribe(in *gjson.Result) (interface{}, error) {
...
+	num, err := svr.parseBlockNumber(blkNum.String())
+	if err != nil {
+		return nil, err
+	}
+	res, err := svr.coreService.BlobSidecarsByHeight(num)
+	switch errors.Cause(err) {
+	case nil:
+		return res, nil
+	case ErrNotFound:
+		return nil, nil
```
Comment (dustinxie) [inline api/web3server.go:1001]: `ErrNotFound` should return nil to caller?

Comment (envestcc) [reply api/web3server.go:1001]: yeah, just like getBlock and getTransaction

File: api/coreservice.go (hunk 4 above)
Comment (envestcc) [reply api/coreservice.go:1243]: I think we can just keep same as [bsc](https://github.com/bnb-chain/bsc/blob/master/core/types/blob_sidecar.go#L25). IMO, this design is to use same structure for getSidercarByBlock and getSidercarByTx

File: api/coreservice.go (hunk 2 above)
Comment (envestcc) [reply api/coreservice.go]: fixed

File: api/coreservice.go (hunk 3 above)
Comment (envestcc) [reply api/coreservice.go]: fixed

**Review by dustinxie — APPROVED** [review]

File: api/coreservice.go [hunk 6]
```diff
@@ -1213,6 +1217,53 @@ func (core *coreService) getBlockByHeight(height uint64) (*apitypes.BlockWithRec
...
+		_, _, index, err := core.ActionByActionHash(txHashes[i])
+		if err != nil {
+			return nil, err
+		}
+		res = append(res, &apitypes.BlobSidecarResult{
+			BlobSidecar: blob,
+			BlockNumber: height,
+			BlockHash:   common.BytesToHash(blkHash[:]),
+			TxIndex:     uint64(index),
+			TxHash:      common.BytesToHash(txHashes[i][:]),
```
Comment (CoderZhi) [inline api/coreservice.go:1241]: do the conversion in line 1230 and save them in local variables.

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**